src/hotspot/share/logging/logFileStreamOutput.cpp

```diff
@@ -146,6 +163,36 @@ int LogFileStreamOutput::write_internal(const LogDecorations& decorations, const
   return written;
 }
 
+int LogFileStreamOutput::write_internal(const LogDecorations& decorations, const char* msg) {
+  int msg_len = checked_cast<int>(strlen(msg));
+
+  // Do not handle multiline messages if foldmultilines has been specified
+  if (_fold_multilines) return write_internal_line(decorations, msg, msg_len);
+
+  // Handle multiline strings: split the string replacing newlines with terminators,
+  // and then force write_internal_line to print all of them (i.e. not stopping at the
+  // first null but until msg_len bytes are printed)
```

Comment:
The string duplication with newline-to-nul conversion likely performs okay for typical short logging statements. But my initial approach would have been to scan through the string looking for the next newline and print it in chunks that way.

Reply:
This might be the better way to go, with `strstr`.

Reply:
I'm not sure if that would be possible. The JVM IO printing functions need a null terminated string, so in the end we would still need to either replace newlines or copy each newline-terminated string to a new one that we would print. In any case, if instead of traversing the string substituting we use `strstr`, under the hood we're still traversing and now I'm not that sure if the approach would be vectorizable

Reply:
I'm still fine with what we have.

Reply:
> The JVM IO printing functions need a null terminated string

No they don't. Instead of using:
```
jio_fprintf(_stream, "%s\n", msg)
```
you use
```
jio_fprintf(_stream, "%*s\n", msg_len, msg)
```
So by advancing a cursor to the new newline you can print all the characters prior, since the last print and so step along the string.

I'm not saying we have to make this change but I do worry that with larger strings the allocation and copying overhead may become an issue.

Reply:
I just pushed some changes with your suggestion. Even though my performance tests didn't show much of an overhead, if we can avoid copying the string it will be better